packages/textfield/src/mwc-textfield.ts

```diff
@@ -0,0 +1,288 @@
+      floatLabel: (shouldFloat) => this.label_.float(shouldFloat),
+      hasLabel: () => !!this.label_,
+      getLabelWidth: () => this.label_.getWidth(),
+      
+      activateLineRipple: () => {
+        if (this.lineRipple_) {
+          this.lineRipple_.activate();
+        }
+      },
+      deactivateLineRipple: () => {
+        if (this.lineRipple_) {
+          this.lineRipple_.deactivate();
+        }
+      },
+      setLineRippleTransformOrigin: (normalizedX) => {
+        if (this.lineRipple_) {
+          this.lineRipple_.setRippleCenter(normalizedX);
+        }
+      },
+
+      notchOutline: (labelWidth, isRtl) => this.outline_.notch(labelWidth, isRtl),
+      closeOutline: () => this.outline_.closeNotch(),
+      hasOutline: () => !!this.outline_,
+
+      registerInputInteractionHandler: (evtType, handler) => this.inputSelector.addEventListener(evtType, handler),
+      deregisterInputInteractionHandler: (evtType, handler) => this.inputSelector.removeEventListener(evtType, handler),
+      getNativeInput: () => this.inputSelector,
+
+      helperText: this.helperText_ ? this.helperText_.foundation : undefined,
+      leadingIcon: this.leadingIcon_ ? this.leadingIcon_.foundation : undefined,
+      trailingIcon: this.trailingIcon_ ? this.trailingIcon_.foundation : undefined
+    };
+  }
+
+  initFactories(
+    lineRippleFactory = (el) => new MDCLineRipple(el),
+    helperTextFactory = (el) => new MDCTextFieldHelperText(el),
+    iconFactory = (el) => new MDCTextFieldIcon(el),
+    labelFactory = (el) => new MDCFloatingLabel(el),
+    outlineFactory = (el) => new MDCNotchedOutline(el)) {
+    
+    const labelElement = this.labelSelector;
+    if (labelElement) {
+      this.label_ = labelFactory(labelElement);
+    }
+    const lineRippleElement = this.lineRippleSelector;
+    if (lineRippleElement) {
+      this.lineRipple_ = lineRippleFactory(lineRippleElement);
+    }
+    const outlineElement = this.outlineSelector;
+    if (outlineElement) {
+      this.outline_ = outlineFactory(outlineElement);
+    }
+    if (this.inputSelector.hasAttribute(ARIA_CONTROLS)) {
+      const helperTextElement = document.getElementById(this.inputSelector.getAttribute(ARIA_CONTROLS) || '');
+      if (helperTextElement) {
+        this.helperText_ = helperTextFactory(helperTextElement);
+      }
+    }
+
+    const iconElements = this.iconSelector;
+    if (iconElements.length > 0) {
+      if (iconElements.length > 1) { // Has both icons.
+        this.leadingIcon_ = iconFactory(iconElements[0]);
+        this.trailingIcon_ = iconFactory(iconElements[1]);
+      } else {
+        if (this.mdcRoot.classList.contains(WITH_LEADING_ICON)) {
+          this.leadingIcon_ = iconFactory(iconElements[0]);
+        } else {
+          this.trailingIcon_ = iconFactory(iconElements[0]);
+        }
+      }
+    }
+
+    //I couldn't get rippleFactory to work. It keeps giving me "unable to access classList of undefined" errors.

```

Comment:
Why you don't use the ripple directive?

Reply:
I was unable to get it to work unfortunately

Reply:
First the import

    import { ripple } from '@material/mwc-ripple/ripple-directive.js';

Then the ripple directive

    <div class="mdc-text-field mdc-text-field--upgraded ${classMap(classes)}" .ripple="${!outlined ? ripple({ unbounded: false }) : undefined}">

Finally you just need to redirect the inputs events to the root element in order to make the ripple work

    <input ... @focus="${this.handleInteractiveEvent}" @blur="${this.handleInteractiveEvent}">

Using something like

    handleInteractiveEvent(evt: MouseEvent) {
        const event = new CustomEvent(evt.type);
        this.mdcRoot.dispatchEvent(event);
    }

Reply:
**RippleOptions** accepts an *interactionNode* but doesn't work as expected because of the lifecycle; the input doesn't exist yet at that point, that's why I would try this approach.

Reply:
I added the changes and it seems to work as expected. I'm not quite sure what "expected" means in the context of this textfield though. It doesn't throw an error, but I see no discernible difference by having the ripple directive versus not having it. I compared Google's own demo [here](https://material-components.github.io/material-components-web-catalog/#/component/text-field), this textfield with the ripple directive, and then without. Do you mind explaining what I should be expecting?